# One stop prpblem:

# Settings

Here are the settings. 

7 states:

    state 0 means now
    
    state 1-5 means time day1 to day5
    
    state 6 means terminal (once we buy one asset, we stop because accoring to Problem2 the buyer buy one asset)
   
6 action:

    action 0-4 means buy asset0-4
    
    action 5 means not buy and go to the next day(also state)
    
    
Reward:
    The benefit of buying each asset is randomly set but has a increasing trend. There is also time value in the reward because in the problem setup, it either is optimizing the execution price or minimizing number of counterparties the buyer talk to.
    
Data instruction:
For each pair of states in (0,6) and actions in (0,5), P[state][action] is a
tuple of the form (probability, nextstate, reward)



In [1]:
import numpy as np
import random

In [35]:
Initial_data={
    0:{0:0,1:0,2:0,3:0,4:0,5:0},
    1:{0:0,1:0,2:0,3:0,4:0,5:0},
    2:{0:0,1:0,2:0,3:0,4:0,5:0},
    3:{0:0,1:0,2:0,3:0,4:0,5:0},
    4:{0:0,1:0,2:0,3:0,4:0,5:0},
    5:{0:0,1:0,2:0,3:0,4:0,5:0},
    6:{0:0,1:0,2:0,3:0,4:0,5:0},
}

In [36]:
def generate_data(data):
    for i in range(0,len(data)-1):
        for j in range(0,len(data[i])-1):
            time_value=5-i
            data[i][j]=(1,6,(random.random()*50*(i+1)/(i+2)+time_value))
        data[i][len(data[i])-1]=(1,i+1,0)
    for i in range(0,len(data[i])):
        data[len(data)-1][i]=(1,6,0)
    return data

In [37]:
data=generate_data(Initial_data)
data

{0: {0: (1, 6, 24.472738431645716),
  1: (1, 6, 29.61647337301111),
  2: (1, 6, 26.672830669303004),
  3: (1, 6, 15.490360700756945),
  4: (1, 6, 25.468591360003312),
  5: (1, 1, 0)},
 1: {0: (1, 6, 19.290863364585064),
  1: (1, 6, 14.854941014404773),
  2: (1, 6, 26.775086288312114),
  3: (1, 6, 4.755199910721819),
  4: (1, 6, 16.267087711610202),
  5: (1, 2, 0)},
 2: {0: (1, 6, 7.671710564653766),
  1: (1, 6, 8.379068812682302),
  2: (1, 6, 40.41933731373332),
  3: (1, 6, 23.521573110172138),
  4: (1, 6, 11.425033091267373),
  5: (1, 3, 0)},
 3: {0: (1, 6, 14.12846515177489),
  1: (1, 6, 31.712676597916754),
  2: (1, 6, 16.597575242961376),
  3: (1, 6, 36.20803427409019),
  4: (1, 6, 30.531822212443807),
  5: (1, 4, 0)},
 4: {0: (1, 6, 13.166553076155543),
  1: (1, 6, 36.55023996413347),
  2: (1, 6, 36.37047096936288),
  3: (1, 6, 26.565637206450692),
  4: (1, 6, 7.67516664675924),
  5: (1, 5, 0)},
 5: {0: (1, 6, 18.235099396579574),
  1: (1, 6, 39.316755312559096),
  2: (1, 6, 32.37

# Using Value_iteration in MDP to solve

In [39]:
def value_iteration(P, nS, nA, gamma=0.9, tol=1e-3):  
    value_function = np.zeros(nS)
    policy = np.zeros(nS, dtype=int)
    while True:
        prev_value_function = np.ones(nS)
        for i in range(nS):
            Qvalue = np.zeros(nA)
            
            for j in range(nA):
                prob, next_state, reward = P[i][j]
                Qvalue[j] = prob * (reward + gamma * value_function[next_state])
            
            prev_value_function[i]=np.max(Qvalue)
            policy[i]=np.argmax(Qvalue)
            
        var=np.linalg.norm(prev_value_function-value_function, ord=np.inf)
        if var>tol:
            value_function=prev_value_function
        else:
            break
    return value_function, policy


In [40]:
value_function,policy=value_iteration(data,len(data),len(data[1]))
print("value_function=",value_function)
print("policy=",policy)


value_function= [32.73966322 36.37740358 40.41933731 36.20803427 36.55023996 39.31675531
  0.        ]
policy= [5 5 2 3 1 1 0]


# Result 

The results shows that the optimal action in day 0 and 1 is waiting, buy asset 2 in day 2, buy asset 3 in day 3 and so on. The optimal value is 40.42 which means buy asset 2 in day 2 is the best option in this setting.